In [1]:
# Dependencies
import pandas as pd

### Transform the zip_to_zcta dataframe

In [2]:
# Import ZIP to ZCTA conversion table. All fields were imported as string to ensure the ZIP Codes retains the "00" prefixes 
zip_to_zcta_df = pd.read_excel("../00_input/zip_to_zcta_2019.xlsx", dtype = 'str')

In [3]:
# Check the ZIP/ZCTA dataframe data, particularly for retention of "00" for ZIP Codes
zip_to_zcta_df.head()

ZIP_CODE     PO_NAME STATE                              ZIP_TYPE   ZCTA  \
0    00501  Holtsville    NY  Post Office or large volume customer  11742   
1    00544  Holtsville    NY  Post Office or large volume customer  11742   
2    00601    Adjuntas    PR                         Zip Code Area  00601   
3    00602      Aguada    PR                         Zip Code Area  00602   
4    00603   Aguadilla    PR                         Zip Code Area  00603   

          zip_join_type  
0  Spatial join to ZCTA  
1  Spatial join to ZCTA  
2      Zip Matches ZCTA  
3      Zip Matches ZCTA  
4      Zip Matches ZCTA

In [4]:
# Review data types in ZIP/ZCTA dataframe
zip_to_zcta_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41107 entries, 0 to 41106
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   ZIP_CODE       41107 non-null  object
 1   PO_NAME        41107 non-null  object
 2   STATE          41106 non-null  object
 3   ZIP_TYPE       41107 non-null  object
 4   ZCTA           41107 non-null  object
 5   zip_join_type  41107 non-null  object
dtypes: object(6)
memory usage: 1.9+ MB


In [5]:
# Check the minimum and maximum length of values in column ZIP_CODE
zip_code_len = zip_to_zcta_df["ZIP_CODE"].apply(len)
print(f"Max zip_code length is {zip_code_len.max()}.")
print(f"Min zip_code length is {zip_code_len.min()}.")

Max zip_code length is 5.
Min zip_code length is 5.


In [6]:
# Check the minimum and maximum length of values in column ZCTA
zcta_len = zip_to_zcta_df["ZCTA"].apply(len)
print(f"Max zcta length is {zcta_len.max()}.")
print(f"Min zcta length is {zcta_len.min()}.")

Max zcta length is 7.
Min zcta length is 5.


In [7]:
# A look into rows with length of values in column ZCTA > 5 
zip_to_zcta_df[zip_to_zcta_df.ZCTA.str.len() > zcta_len.min()]

ZIP_CODE                 PO_NAME STATE  \
41098    96939               Ngerulmud    PW   
41099    96940                   Koror    PW   
41100    96941                 Pohnpei    FM   
41101    96942                   Chuuk    FM   
41102    96943                     Yap    FM   
41103    96944                  Kosrae    FM   
41104    96960                  Majuro    MH   
41105    96970                   Ebeye    MH   
41106    96898  Minor Outlying Islands   NaN   

                                   ZIP_TYPE     ZCTA  \
41098                         Zip Code Area  No ZCTA   
41099  Post Office or large volume customer  No ZCTA   
41100  Post Office or large volume customer  No ZCTA   
41101  Post Office or large volume customer  No ZCTA   
41102  Post Office or large volume customer  No ZCTA   
41103  Post Office or large volume customer  No ZCTA   
41104  Post Office or large volume customer  No ZCTA   
41105  Post Office or large volume customer  No ZCTA   
41106  Post Office or large volume customer  No ZCTA   

                          zip_join_type  
41098  territory zip, no ZCTA available  
41099  territory zip, no ZCTA available  
41100  territory zip, no ZCTA available  
41101  territory zip, no ZCTA available  
41102  territory zip, no ZCTA available  
41103  territory zip, no ZCTA available  
41104  territory zip, no ZCTA available  
41105  territory zip, no ZCTA available  
41106  territory zip, no ZCTA available

In [8]:
# Remove all rows with ZCTA = "No ZCTA"
zip_to_zcta_df = zip_to_zcta_df[~(zip_to_zcta_df.ZCTA.str.len() > zcta_len.min())]

In [9]:
zip_to_zcta_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 41098 entries, 0 to 41097
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   ZIP_CODE       41098 non-null  object
 1   PO_NAME        41098 non-null  object
 2   STATE          41098 non-null  object
 3   ZIP_TYPE       41098 non-null  object
 4   ZCTA           41098 non-null  object
 5   zip_join_type  41098 non-null  object
dtypes: object(6)
memory usage: 2.2+ MB


In [10]:
# Identify any duplicated ZIP Codes as all ZIP Codes should be unique (many ZIP Codes to one ZCTA relationship)
zip_to_zcta_df['ZIP_CODE'].nunique()

41098

In [11]:
zip_to_zcta_df['ZCTA'].nunique()

33166

In [25]:
duplicate_zcta_rows = zip_to_zcta_df[zip_to_zcta_df.duplicated(subset=["ZCTA"], keep=False)].sort_values(by="ZCTA")
duplicate_zcta_rows

ZIP_CODE    PO_NAME STATE                              ZIP_TYPE   ZCTA  \
4        00603  Aguadilla    PR                         Zip Code Area  00603   
5        00604  Aguadilla    PR  Post Office or large volume customer  00603   
6        00605  Aguadilla    PR  Post Office or large volume customer  00603   
10       00612    Arecibo    PR                         Zip Code Area  00612   
11       00613    Arecibo    PR  Post Office or large volume customer  00612   
...        ...        ...   ...                                   ...    ...   
41066    99850     Juneau    AK  Post Office or large volume customer  99801   
41077    99928  Ward Cove    AK  Post Office or large volume customer  99901   
41067    99901  Ketchikan    AK                         Zip Code Area  99901   
41071    99921      Craig    AK                         Zip Code Area  99921   
41079    99950  Ketchikan    AK                         Zip Code Area  99921   

              zip_join_type  
4          Zip Matches ZCTA  
5      Spatial join to ZCTA  
6      Spatial join to ZCTA  
10         Zip Matches ZCTA  
11     Spatial join to ZCTA  
...                     ...  
41066  Spatial join to ZCTA  
41077  Spatial join to ZCTA  
41067      Zip Matches ZCTA  
41071      Zip Matches ZCTA  
41079  Spatial join to ZCTA  

[12878 rows x 6 columns]

### Test merging against zip_code in the zip_code table

In [12]:
# Import cleaned restaurant data
zip_code_df = pd.read_csv("../02_transform_restaurant/zip_code.csv", dtype={'zip_code': 'str'})
zip_code_df.head()

zip_code          city state
0    70301     Thibodaux    LA
1    37863  Pigeon Forge    TN
2    30260        Morrow    GA
3    48204       Detroit    MI
4    48235       Detroit    MI

In [13]:
# An overview of restaurant data
zip_code_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5426 entries, 0 to 5425
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   zip_code  5426 non-null   object
 1   city      5426 non-null   object
 2   state     5426 non-null   object
dtypes: object(3)
memory usage: 127.3+ KB


In [14]:
common_zip_mix = pd.merge(zip_code_df, zip_to_zcta_df, left_on="zip_code", right_on="ZIP_CODE")
common_zip_mix.head()

zip_code          city state ZIP_CODE       PO_NAME STATE       ZIP_TYPE  \
0    70301     Thibodaux    LA    70301     Thibodaux    LA  Zip Code Area   
1    37863  Pigeon Forge    TN    37863  Pigeon Forge    TN  Zip Code Area   
2    30260        Morrow    GA    30260        Morrow    GA  Zip Code Area   
3    48204       Detroit    MI    48204       Detroit    MI  Zip Code Area   
4    48235       Detroit    MI    48235       Detroit    MI  Zip Code Area   

    ZCTA     zip_join_type  
0  70301  Zip Matches ZCTA  
1  37863  Zip Matches ZCTA  
2  30260  Zip Matches ZCTA  
3  48204  Zip Matches ZCTA  
4  48235  Zip Matches ZCTA

In [15]:
common_zip_mix.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5426 entries, 0 to 5425
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   zip_code       5426 non-null   object
 1   city           5426 non-null   object
 2   state          5426 non-null   object
 3   ZIP_CODE       5426 non-null   object
 4   PO_NAME        5426 non-null   object
 5   STATE          5426 non-null   object
 6   ZIP_TYPE       5426 non-null   object
 7   ZCTA           5426 non-null   object
 8   zip_join_type  5426 non-null   object
dtypes: object(9)
memory usage: 423.9+ KB


In [16]:
common_zip_mix[common_zip_mix["city"] != common_zip_mix["PO_NAME"]]

zip_code                   city state ZIP_CODE               PO_NAME  \
112     32435       DeFuniak Springs    FL    32435      Defuniak Springs   
308     32720                 DeLand    FL    32720                Deland   
317     27103          Winston-Salem    NC    27103         Winston Salem   
486     60115                 DeKalb    IL    60115                Dekalb   
620     83814          Coeur d'Alene    ID    83814         Coeur D Alene   
2883    96786               Wahiawā    HI    96786               Wahiawa   
2887    96753                 Kƒ´hei    HI    96753                 Kihei   
3100    70634               DeRidder    LA    70634              Deridder   
3702    27105          Winston-Salem    NC    27105         Winston Salem   
4083    50266        West Des Moines    IA    50266       West des Moines   
4214    87532              Española    NM    87532              Espanola   
4537    83815          Coeur d'Alene    ID    83815         Coeur D Alene   
4711    91011  La Cañada Flintridge    CA    91011  La Canada Flintridge   
5090    27127          Winston-Salem    NC    27127         Winston Salem   
5157    27104          Winston-Salem    NC    27104         Winston Salem   
5424    27106          Winston-Salem    NC    27106         Winston Salem   
5425    27107          Winston-Salem    NC    27107         Winston Salem   

     STATE       ZIP_TYPE   ZCTA     zip_join_type  
112     FL  Zip Code Area  32435  Zip Matches ZCTA  
308     FL  Zip Code Area  32720  Zip Matches ZCTA  
317     NC  Zip Code Area  27103  Zip Matches ZCTA  
486     IL  Zip Code Area  60115  Zip Matches ZCTA  
620     ID  Zip Code Area  83814  Zip Matches ZCTA  
2883    HI  Zip Code Area  96786  Zip Matches ZCTA  
2887    HI  Zip Code Area  96753  Zip Matches ZCTA  
3100    LA  Zip Code Area  70634  Zip Matches ZCTA  
3702    NC  Zip Code Area  27105  Zip Matches ZCTA  
4083    IA  Zip Code Area  50266  Zip Matches ZCTA  
4214    NM  Zip Code Area  87532  Zip Matches ZCTA  
4537    ID  Zip Code Area  83815  Zip Matches ZCTA  
4711    CA  Zip Code Area  91011  Zip Matches ZCTA  
5090    NC  Zip Code Area  27127  Zip Matches ZCTA  
5157    NC  Zip Code Area  27104  Zip Matches ZCTA  
5424    NC  Zip Code Area  27106  Zip Matches ZCTA  
5425    NC  Zip Code Area  27107  Zip Matches ZCTA

As can be seen from above, most of the "city" names in the zip_code table are actually the same as the "PO_NAME" names in zip_to_zcta table. Even with some differences, a closer look tells us that they are also the same.

As all zip_code in zip_code table are in zip_to_zcta table, we can just use zip_to_zcta table to map zip_code in restaurant address table to ZCTA code in the census table.

In [17]:
zip_to_zcta_df["ZIP_TYPE"].unique()

array(['Post Office or large volume customer', 'Zip Code Area',
       'populated ZCTA, missing zip', 'Territory ZCTA Add'], dtype=object)

In [18]:
zip_to_zcta_df["zip_join_type"].unique()

array(['Spatial join to ZCTA', 'Zip Matches ZCTA',
       'populated ZCTA, missing zip', 'Territory ZCTA Add'], dtype=object)

In [20]:
# Remove unnecessary columns to refine dataset
final_zip_to_zcta = zip_to_zcta_df[["ZIP_CODE", "PO_NAME", "STATE", "ZCTA"]]
final_zip_to_zcta.head()

ZIP_CODE     PO_NAME STATE   ZCTA
0    00501  Holtsville    NY  11742
1    00544  Holtsville    NY  11742
2    00601    Adjuntas    PR  00601
3    00602      Aguada    PR  00602
4    00603   Aguadilla    PR  00603

In [21]:
# Change ZIP_CODE into zip_code
clean_zip_to_zcta = final_zip_to_zcta.rename(columns={"ZIP_CODE": "zip_code", "PO_NAME": "city_or_po_name", "STATE": "state"})
clean_zip_to_zcta.head()

zip_code city_or_po_name state   ZCTA
0    00501      Holtsville    NY  11742
1    00544      Holtsville    NY  11742
2    00601        Adjuntas    PR  00601
3    00602          Aguada    PR  00602
4    00603       Aguadilla    PR  00603

In [22]:
clean_zip_to_zcta.to_csv("clean_zip_to_zcta.csv", index=False)

### Test merging against ZCTA in the census table

In [26]:
census_df = pd.read_csv("../02_transform_census/clean_census.csv", dtype={'ZCTA': 'str'})
census_df.head()

ZCTA  population  median_age  median_household_income  per_capita_income  \
0  00601       17242        40.5                  13092.0             6999.0   
1  00602       38442        42.3                  16358.0             9277.0   
2  00603       48814        41.1                  16603.0            11307.0   
3  00606        6437        43.3                  12832.0             5943.0   
4  00610       27073        42.1                  19309.0            10220.0   

   poverty_count  unemployment_count  
0          10772                2316  
1          19611                1927  
2          24337                3124  
3           4163                 230  
4          11724                1290

In [27]:
census_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32458 entries, 0 to 32457
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ZCTA                     32458 non-null  object 
 1   population               32458 non-null  int64  
 2   median_age               32458 non-null  float64
 3   median_household_income  30891 non-null  float64
 4   per_capita_income        32455 non-null  float64
 5   poverty_count            32458 non-null  int64  
 6   unemployment_count       32458 non-null  int64  
dtypes: float64(3), int64(3), object(1)
memory usage: 1.7+ MB


In [28]:
zcta_census = pd.merge(census_df, zip_to_zcta_df, how="outer", on="ZCTA")
zcta_census.head()

ZCTA  population  median_age  median_household_income  per_capita_income  \
0  00601     17242.0        40.5                  13092.0             6999.0   
1  00602     38442.0        42.3                  16358.0             9277.0   
2  00603     48814.0        41.1                  16603.0            11307.0   
3  00603     48814.0        41.1                  16603.0            11307.0   
4  00603     48814.0        41.1                  16603.0            11307.0   

   poverty_count  unemployment_count ZIP_CODE    PO_NAME STATE  \
0        10772.0              2316.0    00601   Adjuntas    PR   
1        19611.0              1927.0    00602     Aguada    PR   
2        24337.0              3124.0    00603  Aguadilla    PR   
3        24337.0              3124.0    00604  Aguadilla    PR   
4        24337.0              3124.0    00605  Aguadilla    PR   

                               ZIP_TYPE         zip_join_type  
0                         Zip Code Area      Zip Matches ZCTA  
1                         Zip Code Area      Zip Matches ZCTA  
2                         Zip Code Area      Zip Matches ZCTA  
3  Post Office or large volume customer  Spatial join to ZCTA  
4  Post Office or large volume customer  Spatial join to ZCTA

In [29]:
zcta_census.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 41098 entries, 0 to 41097
Data columns (total 12 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ZCTA                     41098 non-null  object 
 1   population               40296 non-null  float64
 2   median_age               40296 non-null  float64
 3   median_household_income  38669 non-null  float64
 4   per_capita_income        40293 non-null  float64
 5   poverty_count            40296 non-null  float64
 6   unemployment_count       40296 non-null  float64
 7   ZIP_CODE                 41098 non-null  object 
 8   PO_NAME                  41098 non-null  object 
 9   STATE                    41098 non-null  object 
 10  ZIP_TYPE                 41098 non-null  object 
 11  zip_join_type            41098 non-null  object 
dtypes: float64(6), object(6)
memory usage: 4.1+ MB


Comparing the number of non-null ZCTA (41099) and non-null zip_code (41098), there is one ZCTA in the census table without a matching ZCTA in the zip_to_zcta table.

In [30]:
zcta_census[zcta_census["zip_code"].isnull()]

KeyError: 'zip_code'

In [ ]:
zcta_to_remove = zcta_census[zcta_census["zip_code"].isnull()]["ZCTA"].to_list()
zcta_to_remove

In [ ]:
census_df = census_df[~census_df["ZCTA"].isin(zcta_to_remove)]

In [ ]:
census_df.info()

In [ ]:
# Update to census csv
census_df.to_csv('../02_transform_census/clean_census.csv', index=False)

### Export cleaned zip_to_zcta_df to csv

In [ ]:
# Export ZIP_to_ZCTA dataframe into CSV file
zip_to_zcta_df.to_csv('zcta.csv', index=False)